### ISSUE TITLE
Request for torch.distributions primitives

### ISSUE THREAD

**Issue by jerry73204** [LaurentMazare/tch-rs#29]
It seems current tch-rs lacks distribution primitives like [torch.distributions](https://pytorch.org/docs/stable/distributions.html) and [tf.distributions](https://www.tensorflow.org/api_docs/python/tf/distributions). Though it already have probability generators, there's still rooms for probability arithmetic and deductions.

I see a promising [crate rv](https://docs.rs/crate/rv/0.2.2) that may fit my needs. However, its interface is not general enough. We cannot pass a mean/std tensor to rv's Gaussion new() function.

Porting torch.distributions may solve the problem. It requires some patience to make it work, or I could go ahead to improve rv. I'd like to know if author has any plan to port the module, or leaves it to other crates.

If you're asking which place needs this feature, I'm implementing ELBO loss for GQN ([paper](https://deepmind.com/documents/211/Neural_Scene_Representation_and_Rendering_preprint.pdf)).

**Comment by LaurentMazare** [comment]
Thanks for the suggestion. This indeed seems like a nice thing to add - the underlying torch primitives for sampling should already exist in the rust wrappers, e.g. [normal](https://docs.rs/tch/0.0.8/tch/struct.Tensor.html#method.normal) so I guess this would mostly consist in adding some trait for distributions and implementing it mimicking the python implementation, e.g. [Normal](https://pytorch.org/docs/stable/_modules/torch/distributions/normal.html#Normal).
Would that be useful to you?
Also which methods/distributions would be the most interesting to you?

**Comment by jerry73204** [comment]
Yes, I see that normal function. I need probability arithmetic rather than prob generators. Take the GQN for example, I need a Normal object which parameters are defined by tensors, and compute the log_prob (log prob mass function) for any value on the Normal. I also have to compute KL div for two normals. There's no need for sampling.

[Torch](https://pytorch.org/docs/stable/distributions.html#torch.distributions.distribution.Distribution.log_prob) and [TensorFlow](https://www.tensorflow.org/api_docs/python/tf/distributions/Normal) have their own such function. Interestingly, you'll see NotImplementedError if you look into Torch's source code. So I bet improving rv would be a good direction. Currently I just write raw formulas with cautions on floating point precisions.

**Comment by LaurentMazare** [comment]
Do you need to backprop through your KL divergence ? If that's the case I'm not sure that rv could be used out of the box but maybe I'm missing something ?
When looking at the ['Normal' distribution in torch](https://pytorch.org/docs/stable/_modules/torch/distributions/normal.html#Normal), I don't see much NotImplementedError besides in the base Distribution class. Also the kl divergence for the normal distribution can be found [here](https://pytorch.org/docs/stable/_modules/torch/distributions/kl.html#kl_divergence).

**Comment by jerry73204** [comment]
The NotImplememtedError goes [here](https://pytorch.org/docs/stable/_modules/torch/distributions/distribution.html#Distribution.log_prob). Sorry for inprecise comment. 

Backprop is desired in my case. As far as I know, torch.distributioms looks like an add-on of pytorch rather than in libtorch. Implementation in Rust is like a completely new library. 

Let me write my some code for this and to see any proper way to build this feature.

**Comment by LaurentMazare** [comment]
Re NotImplementedError, indeed that's the base class for distributions in pytorch so nothing is implemented here and the various methods are implemented in the derived classes like 'Normal'.
The distribution bit is included within the main pytorch repo and package (contrary to say vision which is an external repo and pypi package). I don't have a proper opinion on where this belongs - in the main crate or in a separate one but starting by an external crate sounds good and if there is some upside to merge it in the main repo we can consider it later.
Let me know if you notice some pytorch primitives missing from tch-rs that could be useful to you!

**Comment by LaurentMazare** [comment]
Just to mention that I added a variational auto-encoder example to tch-rs. This inctludes a KL divergence loss [here](https://github.com/LaurentMazare/tch-rs/blob/master/examples/vae/main.rs#L60).
It's certainly very far away from what a nice `distributions` api would provide but it may be handy.

**Comment by vegapit** [comment]
It is just a case of adding numerical approximation functions for pdfs and cdfs of popular statistical distributions. I personally just implemented the Normal cdf derived in this [article](https://www.hindawi.com/journals/mpe/2012/124029/) using tensors and could normally calculate the gradients with the library:

```
fn norm_cdf(x: &Tensor) -> Tensor {
    let denom = ((-358.0 * x / 23.0) + 111.0 * (37.0 * x / 294.0).atan()).exp() + 1.0;
    1.0 / denom
}
```

**Comment by LaurentMazare** [comment]
@vegapit yes that's mostly about adding such functions and probably some traits for the various distributions.
You can see the implementation for the normal distribution in the python api [here](https://pytorch.org/docs/stable/_modules/torch/distributions/normal.html#Normal). The code for the cdf is a bit different from yours and relies on `torch.erf`. Not sure which one has the best precision.

```python
def cdf(self, value):
        return 0.5 * (1 + torch.erf((value - self.loc) * self.scale.reciprocal() / math.sqrt(2)))
```

**Comment by vegapit** [comment]
@LaurentMazare I did not know you actually had already added the error function implementation in Torch, otherwise I would have used it. I do not know what the numerical approximation in `torch.erf` is but my guess is that its precision must be similar to the function I described above.

**Comment by LaurentMazare** [comment]
Btw the error function is also available in tch-rs https://docs.rs/tch/0.1.0/tch/struct.Tensor.html#method.erf
(as the rust low level wrappers are automatically generated, we mostly get these for free)

**Comment by jerry73204** [comment]
A little update here. In my previous gqnrs project has some useful [traits](https://github.com/jerry73204/gqnrs/blob/master/src/dist/mod.rs) for prob distributions. There is only Normal dist there. Suppose we can start a working branch to fill the blanks for {Bernoulli,Exponential,Categorical, etc} dists?

**Comment by LaurentMazare** [comment]
Yes that kind of trait would indeed probably be useful, @vegapit do you think this would cover your use case?

**Comment by vegapit** [comment]
I use Torch to solve for maximum likelihood in non-linear parametrised models. The density functions can be reconstructed or approximated using tensor methods but I guess it is clearly more user friendly to provide them in wrappers as Pytorch does

**Comment by spebern** [comment]
I started to work on a crate to port the distributions: https://github.com/spebern/tch-distr-rs
Besides porting, the most tedious work is to test everything.

I think the best way to ease porting is to test against the python implementations directly by supplying the same input and comparing the outputs. Later, this can also be extended with fuzzy input.

The `Distribution` trait is open for discussion and pull requests are more than welcome to add more distributions/tests.

**Comment by LaurentMazare** [comment]
This looks very nice, thanks for sharing! Once this has been polished/tested a bit, we could probably mention it in the `tch-rs` main readme file for better discoverability if that's ok with you.

**Comment by spebern** [comment]
That would be really nice! It definitely needs some polishing and more implemented distributions, but
I really think that the testing against python implementations takes away a lot of work.

**Comment by dbsxdbsx** [comment]
> This looks very nice, thanks for sharing! Once this has been polished/tested a bit, we could probably mention it in the `tch-rs` main readme file for better discoverability if that's ok with you.

@LaurentMazare, I wonder if it a good idea to introduce [tch-distr-rs](https://github.com/spebern/tch-distr-rs)  of @spebern  as a feature for torch-rs. So that user no need to introduce 2 crates when using torch in rust.

**Comment by LaurentMazare** [comment]
If it's just to avoid having an additional dependency for crates that would want to use this, I would lean more towards keeping an external crate, and in general having smaller composable crates for the bits that are outside of the core `tch-rs`, e.g. I'm more thinking about moving the vision models in their own crate, the RL bits to their own thing too etc.
